Add scroll text widget

Bauthe opened the pull request:
Adds a widget that can scroll text.
Here is a config to showcase this widget: https://gist.github.com/Bauthe/258371d2734f6d5d9f9ae38ad5cc3617

ramnes commented:
Interesting! But please:
* remove all the noise related to reformatting
* add a widget (or modify an existing one) so that this base class is actually used somewhere and do not just end up being dead code
* ideally, add tests

m-col commented:
General question: would it make sense and be possible to add the scrolling functionality to the existing `_TextBox` or even `_Widget` base classes?

Bauthe commented:
@ramnes
> * remove all the noise related to reformatting

Done

> * add a widget (or modify an existing one) so that this base class is actually used somewhere and do not just end up being dead code

I modified the Mpris2 widget in a second commit. What I've added:
- Smooth text scrolling
- Control of the player through a dbus interface
- More formatting options

> * ideally, add tests

I can help with that when I have some feedback about the functionnality


@m-col 
> General question: would it make sense and be possible to add the scrolling functionality to the existing `_TextBox` or even `_Widget` base classes?

It is possible on `_TextBox`, but I don't think we want it because it adds a lot of overhead/clutter to the drawing function. I would suggest to keep it separate for readability/performance.

Bauthe commented:
@elParaguayo @tych0 mentioning you because you were interested in my first PR #1927

elParaguayo commented:
Thanks. I was going to make some changes to this widget myself as I hated the metadata/playback handling section of the code. However, I can see you're way ahead of me which is great!

I also like the fact that you've added some basic controls - (shameless plug alert) that's what I had in mind as part of #2251.

I'll take a look later this week and let you know if I have any comments.

elParaguayo commented:
The scrolling looks really nice and a big improvement on the character scrolling.

A few comments on this:
1) Is MarkupSafe really needed as a dependency?
2) There are a couple of glitches but I haven't dug into the code to see what's causing them:

i) When scrolling there's a small artefact on both the left and right sides, like those pixels aren't part of the scroll.

ii) Sometimes the widget changes size (could due to some empty characters at the end of the string). This results in the widget being too long to start with but resetting itself when the scroll loop restarts.

However, metadata, playback status and controls all seem to work.

tych0 commented on libqtile/widget/mpris2widget.py:
```diff
@@ -22,158 +22,152 @@
+        ('play_char', '', '"status_char" to display a song is playing.'),
+        ('pause_char', '', '"status_char" to display when a song is paused.'),
+        ('txt_inactive', '', 'Text to display when the player is inactive'),
+        ('scrolling_enabled', True, 'Wether to scroll the text.'),
```
This looks like a bunch of breaking changes. Please either 1. implement a migration for these changes via qtile-migrate or 2. figure out how to make them non-breaking.

Bauthe replied:
I've adapted the signature so that it fits as much as possible the `Mpd2` widget syntax. I will take a look at `qtile migrate`, it seems awesome.

tych0 replied:
The problem is that people who are existing users of the mpris widget will not necessarily be familiar with the mpd2 widget. It is best to make changes non-breaking if possible.

Bauthe replied:
Yes I get it. I will try to keep the old names when it can still make sense. However, what should I do to deprecate certain parameters smoothly ? Is there a guideline ?

tych0 replied:
Yes, you'll need to implement a `qtile migrate` migration for anything you'd like to change.

Bauthe commented:
@elParaguayo thanks for the input
> 1. Is MarkupSafe really needed as a dependency?

Not at all, after looking it up it seems feasible to escape them manually: https://stackoverflow.com/questions/1091945/what-characters-do-i-need-to-escape-in-xml-documents

> i) When scrolling there's a small artefact on both the left and right sides, like those pixels aren't part of the scroll.
>
> ii) Sometimes the widget changes size (could due to some empty characters at the end of the string). This results in the widget being too long to start with but resetting itself when the scroll loop restarts.

I don't encounter these issues, so I don't see how to solve them. If you use a particular config, could you share it ? Or send a gif of these glitchy behaviours ? Here is how it behaves on my side:

![MPRIS2widget](https://user-images.githubusercontent.com/49082598/110047823-d9cddb00-7d4e-11eb-96c7-ec2973bbf479.gif)

tych0 commented on libqtile/widget/mpris2widget.py:
```diff
@@ -22,158 +22,154 @@
-        self.text = ''
-        self.bar.draw()
-
-    def cmd_info(self):
```
This is a public API, so we should avoid dropping it.

tych0 commented on libqtile/widget/mpris2widget.py:
```diff
@@ -22,158 +22,154 @@
-                        bus_name=self.objname,
-                        path="/org/mpris/MediaPlayer2",
-                        dbus_interface="org.freedesktop.DBus.Properties")
+            self.message,
```
This looks like a bunch of whitespace changes; can we drop these whitespace changes (and all others, e.g. the `"".format() -> f""`) from this patch?

elParaguayo commented:
You can see both issues in the gif below:
![mpris](https://user-images.githubusercontent.com/946265/110164780-c3289200-7de9-11eb-83f6-4b7b5a22625c.gif)


1) Artefacts at either end of the scroll
2) After stopping and starting, the widget has moved to the left but resets correctly after the first scroll. (This is when you see the pause symbol, then the text disappears as I stop the track. When I restart the track, the text is in the wrong place).

Bauthe commented:
I've managed to reproduce the "widget moving to left behaviour" by putting a `bar.STRETCH` spacer before my widget, so I might be able to debug it easily. I think it is because I don't redraw the bar everytime I should. For the artefacts on each end, I can't reproduce it, but I think it is due to imprecisions in floating point values, e.g. in `_ScrollText.draw`: `padding = int(min(self.actual_padding, self.width / 2))`.

elParaguayo commented:
If I set `padding` to 0 in that file the artefacts disappear so you are right about that being the cause.

I'd like to take a look at the `_ScrollText` class some more as I'm not sure I follow all the logic at the moment.

m-col commented:
> It is possible on _TextBox, but I don't think we want it because it adds a lot of overhead/clutter to the drawing function. I would suggest to keep it separate for readability/performance.

To me it seems an arbitrary decision for some text-based widgets to be scollable and others not. What kind of performance hit does a scrollable widget really experience, even when not scrolling? I foresee once there is one scrollable widget people will want to scroll other ones. Indeed I have a few text-based widgets I'd like this feature on. Can we work around any expected performance hit when scrolling is not enabled?
